[TAN-447] Save phase reports to reports table

alexander-cit opened the pull request:
~~Inspired by polls. We save them in a common endpoint by specifying participation context type and id~~ https://github.com/CitizenLabDotCo/citizenlab/blob/f8d97a3fbac0b1d6804dc59127a125852359df0d/front/app/api/poll_questions/useAddPollQuestion.ts#L21-L24

alexander-cit committed:
[TAN-447] Save phase reports to reports table

alexander-cit committed:
[TAN-447] Add specs for report context create and get actions

alexander-cit committed:
[TAN-447] Prevent n+1 SQL queries

alexander-cit committed:
[TAN-447] Adjust specs to new requirements

alexander-cit commented on back/engines/commercial/report_builder/app/models/report_builder/report.rb:
```diff
@@ -29,6 +32,8 @@ class Report < ::ApplicationRecord
     accepts_nested_attributes_for :layout
 
-    validates :name, presence: true, uniqueness: true
+    scope :with_platform_context, -> { where(participation_context_type: nil) }
```
@luucvanderzee It's how I call the reports w/o any participation context. Maybe you'll need this name too. I'm also open to changing it if you think it's unclear.

luucvanderzee replied:
Makes sense, I don't see why we would need a name for these type of reports

alexander-cit replied:
I wasn't super clear. With my comment, I meant that I call reports that do not belong to phases (or any other participation context) by the name of `Reports with (entire) platform participation context`.

alexander-cit commented on back/engines/commercial/report_builder/app/controllers/report_builder/web_api/v1/reports_controller.rb:
```diff
@@ -61,14 +61,16 @@ def report
 
         def create_params
-          @create_params ||= {
+          context_params = params.require(:report).permit(:participation_context_type, :participation_context_id)
```
@luucvanderzee To create a report connected to a phase, you can add two new params: `participation_context_type` and `participation_context_id`. `name` is not required anymore (the doc will be updated soon) https://developers.citizenlab.co/api-docs/ee/frontweb_api/TAN-447-save-phase-reports/reports/create_a_phase_report.html

alexander-cit commented on back/app/serializers/web_api/v1/phase_serializer.rb:
```diff
@@ -15,6 +15,8 @@ class WebApi::V1::PhaseSerializer < WebApi::V1::ParticipationContextSerializer
     user_basket object, params
   end
 
+  has_one :report, serializer: ReportBuilder::WebApi::V1::ReportSerializer
```
@luucvanderzee The report data will be included in the phase serializer https://developers.citizenlab.co/api-docs/ee/frontweb_api/TAN-447-save-phase-reports/phases/get_one_phase_by_id.html

luucvanderzee replied:
I can see it in the relationships, is that what you mean? I don't see it in the `included` array

alexander-cit replied:
> I can see it in the relationships, is that what you mean?

Yes, exactly.

> I don't see it in the included array

`included` array is used to avoid fetching this "included" data with a second HTTP call. Because "included" data is included in the first HTTP call.

But for now, I believe we don't need it as reports will be rarely fetched. Or at least only on pages for only one phase. In this case, it's fine to make a second HTTP call.

luucvanderzee reviewed (commented):
Looks good, just one question

alexander-cit committed:
Merge branch 'master' into TAN-447-save-phase-reports

alexander-cit committed:
[TAN-447] Replace participation_context ref with phase_id

1. We're going to get rid of participation_context
2. KISS
3. W/o a polymorphic association, we can use foreign keys

luucvanderzee committed:
Merge branch 'master' into TAN-453-add-new-report-builder

luucvanderzee committed:
Add phase_id param to useAddReport

luucvanderzee committed:
Remove injectIntl and resources from participationContext

luucvanderzee committed:
Button to create report (WIP)

luucvanderzee committed:
Add report relationship to phase

alexander-cit committed:
[TAN-447] Add report association to phase responses

alexander-cit committed:
[TAN-447] Add phase association to report

luucvanderzee committed:
Add report row in phase setup

luucvanderzee committed:
Merge branch 'TAN-447-save-phase-reports' into TAN-453-add-new-report-builder

alexander-cit committed:
[TAN-447] Add phase_reports feature

alexander-cit committed:
[TAN-447] Enable phase_reports in dev env

alexander-cit committed:
Merge branch 'master' into TAN-447-save-phase-reports

luucvanderzee committed:
_

luucvanderzee committed:
Merge branch 'TAN-447-save-phase-reports' into TAN-453-add-new-report-builder

Commit:
Translations updated by CI (extract-intl)

alexander-cit committed:
[TAN-447] Fix specs by making phase optional for report

alexander-cit committed:
[TAN-447] Improve some constant names

alexander-cit committed:
[TAN-447] Add phase data to report acceptance specs

luucvanderzee committed:
Add feature flag and custom copy

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
Remove empty error handling

luucvanderzee committed:
Merge pull request #6384 from CitizenLabDotCo/TAN-453-add-new-report-builder

TAN-453 Add new report builder

alexander-cit committed:
Merge branch 'master' into TAN-447-save-phase-reports
